File: 4 Binary Search/1 Binary Search (Easy)/Solution.cpp
#include <algorithm>
#include <vector>
#include <iostream>
#include <iterator>

using namespace std;


class Solution {
public:
    static int search(vector<int>& nums, int target) {
        long left = 0;
        long right = nums.size();

        while (left <= right) {
            long* center = (long*) malloc(sizeof(long));
            *center = left + (right - left) / 2;
            if (nums[*center] == target) {
                return *center;
            }
            if (nums[*center] < target) {
                left = *center + 1;
            }
            else {
                right = *center - 1;
            }
            free(center);
        }

        return -1;
    }
};

int main() {
    std::vector<int> vec = {5};
    cout << "Hello\n";
    cout << Solution::search(vec, -5) << endl;
    return 0;
}

File: 4 Binary Search/1 Binary Search (Easy)/Solution.py
def search(nums: list[int], target: int) -> int:
        center = len(nums) // 2
        left_count = 0
        
        while nums[center] != target :
            if (len(nums) == 1) :
                return -1
            elif nums[center] > target :
                nums = nums[:center]
            else :
                nums = nums[center:]
                left_count += center
            center = len(nums) // 2
            
        return center + left_count
    
if __name__ == "__main__" :
    print(search([2, 3, 5, 6, 7, 12, 44, 56, 57], 12))